src/pseudo/PspUpf.jl

```diff
@@ -190,6 +214,42 @@ function eval_psp_local_fourier(psp::PspUpf, q::T)::T where {T <: Real}
     4T(π) * (s - psp.Zion / q) / q
 end
 
+function eval_psp_rho_valence_real(psp::PspUpf, r::T) where {T <: Real}
+    psp.r2_4π_ρion_interp(r) / (r^2 * 4T(π))
+end
+
+"""
+    eval_psp_rho_valence_fourier(psp::PspUpf, q<:Real)
+
+For UPFs, the integral is transformed into the following sum:
+Σ{i} j_0(q r[i]) r^2 4π ρval[i] dr[i]
+"""
+function eval_psp_rho_valence_fourier(psp::PspUpf, q::T) where {T <: Real}
+    s = zero(T)
+    @inbounds for ir = eachindex(psp.r2_4π_ρion_dr)
+        s += sphericalbesselj_fast(0, q * psp.rgrid[ir]) * psp.r2_4π_ρion_dr[ir]
+    end
+    s
```

Comment:
this pattern pops up a lot. We should ahve a function for it.

Reply:
In general this pattern, the hash map form-factor calculation / memoization, and structure factor calculation all show up a lot. I think QE calculates the structure factors once then stores them because they're used so often; could also be a possibility.

Reply:
I don't have a strong opinion on that. I think if the performance is ok we can also just leave it as is and make a comment we should refactor this later. In line with that my suggestion to have a function for this pattern makes it a little easier to find the places where the repeated calculation of the bessel form factors terms happen.

Reply:
could just be a `bessel_transform(l, q, rgrid, fgrid)`

Reply:
ok, but please open an issue to collect refactoring that is needed once PseudoPotentialIO is in such that we don't forget.